wen-restart/src/wen_restart.rs

```diff
@@ -341,6 +360,75 @@ pub(crate) fn find_heaviest_fork(
     Ok((heaviest_fork_slot, heaviest_fork_bankhash))
 }
 
+// Given the agreed upon slot, make corresponding bank root and generate snapshot.
+pub(crate) fn generate_snapshot(
+    bank_forks: Arc<RwLock<BankForks>>,
+    snapshot_config: &SnapshotConfig,
+    accounts_background_request_sender: &AbsRequestSender,
+    genesis_config_hash: Hash,
+    new_root_slot: Slot,
+    exit: &Arc<AtomicBool>,
+) -> Result<GenerateSnapshotRecord> {
+    let new_root_bank;
+    {
+        let mut my_bank_forks = bank_forks.write().unwrap();
+        let old_root_bank = my_bank_forks.root_bank();
+        old_root_bank.register_hard_fork(new_root_slot);
+        new_root_bank = my_bank_forks.get(new_root_slot).unwrap();
+        new_root_bank.rehash();
+        // Only root banks can have snapshots generated.
+        my_bank_forks.set_root(
+            new_root_slot,
+            accounts_background_request_sender,
+            Some(new_root_slot),
+        )?;
```

Comment:
All options feel like they peek too much into snapshot logic so they really don't belong to the current PR, but I guess my use case is unique.

I would say the method feeling more natural is to ensure there are no concurrent snapshot requests at the lower level. Maybe someday we changed the program logic and set_root() got called on two slots eligible for snapshot consecutively, the library need to be robust at fending off errors no matter how stupid the callers are.

But I can also live with the assumption of smart callers and ask the caller to check current status as well.

Let me know which option you prefer and whether you want that in a separate PR.

Reply:
It is manual, operators have to do a validator restart with --wen_restart command line option.

When they enter the restart there shouldn't be any new blocks confirmed, so there should be no `set_root() `happening automatically. But I think we were worried about the `set_root()` call before  `bank_to_incremental_snapshot_archive()` triggering a snapshot request?

Reply:
Did 3) and fixed the incremental/full snapshot generation, see if you like this approach?

Reply:
Oops, you are right, I meant to say option 1) when I said option 3). Since wen_restart is used at most once per quarter, I want to mostly piggyback on existing code which will be used more often. So 1) seems a bit better than 3).

Thanks for handling the refactoring.

Reply:
Changed to always use full snapshot for now and using wen_restart subdir under the full snapshot dir. How does this look now?

Reply:
> (3) We could put code in here to replace set_root() that checks if an EAH is needed, and then calls all the same functions as ABS and AHV to compute the EAH. This means it would all be in the foreground. We'd also be able to use all the available cores on the machine to make the EAH calculation even faster.

Seems like for option 3 we should refactor ABS to make this a utility function that is shared so that we can maintain it